src/integration/resources/docker/dockerexternal/etcdintegration/cluster.go

```diff
@@ -0,0 +1,293 @@
+// the Blackhole method.
+// Bridge will only be active if cfg.UseBridge is true; calling this method otherwise will panic.
+func (n *Node) Bridge() *bridge.Bridge {
+	if !n.cfg.UseBridge {
+		panic("EtcdNode wasn't configured to use a Bridge; pass EtcdClusterUseBridge(true) to enable.")
+	}
+	return n.bridge
+}
+
+// Restart starts a stopped etcd container, stopping it first if it's not already.
+func (n *Node) Restart(t testingT) error {
+	ctx, cancel := context.WithTimeout(context.Background(), startTimeout)
+	defer cancel()
+	requireNoError(t, n.resource.Restart(ctx))
+	return nil
+}
+
+// start starts the etcd node. It is private because it isn't part of the etcd/integration package API, and
+// should only be called by Cluster.start.
+func (n *Node) start(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, startTimeout)
+	defer cancel()
+
+	if err := n.resource.Setup(ctx); err != nil {
+		return err
+	}
+
+	address := n.resource.Address()
+	if n.cfg.UseBridge {
+		addr, err := n.setupBridge()
+		if err != nil {
+			return fmt.Errorf("setting up connection bridge for etcd node: %w", err)
+		}
+		address = addr
+	}
+
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{"http://" + address},
+		DialOptions: []grpc.DialOption{grpc.WithBlock()},
+		DialTimeout: 5 * time.Second,
+		Logger:      n.logger,
+	})
+
+	if err != nil {
+		return fmt.Errorf("constructing etcd client for member: %w", err)
+	}
+
+	n.logger.Info("Connecting to docker etcd using host machine port",
+		zap.String("endpoint", address),
+	)
+
+	n.Client = etcdCli
+	return nil
+}
+
+// setupBridge puts a man-in-the-middle listener in between the etcd docker process and the client. See Bridge() for
+// details.
+// Returns the new address of the bridge, which clients should connect to.
+func (n *Node) setupBridge() (string, error) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return "", fmt.Errorf("setting up listener for bridge: %w", err)
+	}
+
+	n.logger.Info("etcd bridge is listening", zap.String("addr", listener.Addr().String()))
+
+	// dialer = make connections to the etcd container
+	// listener = the bridge's inbounds
+	n.bridge, err = bridge.New(dialer{hostport: n.resource.Address()}, listener)
+	if err != nil {
+		return "", err
+	}
+
+	return listener.Addr().String(), nil
+}
+
+func (n *Node) close(ctx context.Context) error {
+	var err xerrors.MultiError
+	err = err.Add(n.Client.Close())
+	return err.Add(n.resource.Close(ctx)).FinalError()
+}
+
+type dialer struct {
+	hostport string
+}
+
+func (d dialer) Dial() (net.Conn, error) {
+	return net.Dial("tcp", d.hostport)
+}
+
+// testingT wraps *testing.T. Allows us to not directly depend on *testing package.
+type testingT interface {
+	zaptest.TestingT
+
+	Cleanup(func())
+}
+
+// require.NoError, but without depending on require package
```

Comment:
Curious why, since we use this package elsewhere?

Reply:
Basically this allows us to avoid pulling in the testing package to our "production" codebase--since these aren't *test.go files, they shouldn't depend on *testing.T. More of a hygiene thing than anything else.

Reply:
... nvm that's for testingT. I'm actually not sure; undoing it.